collectors/diskspace.plugin/plugin_diskspace.c

```diff
@@ -106,10 +123,208 @@ int mount_point_is_protected(char *mount_point)
     return 0;
 }
 
-static inline void do_disk_space_stats(struct mountinfo *mi, int update_every) {
+// a copy of basic mountinfo fields
+struct basic_mountinfo {
+    char *persistent_id;    
+    char *root;             
+    char *mount_point;      
+    char *filesystem;       
+
+    struct basic_mountinfo *next;
+};
+
+static struct basic_mountinfo *slow_mountinfo_tmp_root = NULL;
+static netdata_mutex_t slow_mountinfo_mutex;
+
+static struct basic_mountinfo *basic_mountinfo_create_and_copy(struct mountinfo* mi)
+{
+    struct basic_mountinfo *bmi = callocz(1, sizeof(struct basic_mountinfo));
+    
+    if (mi) {
+        bmi->persistent_id = strdupz(mi->persistent_id);
+        bmi->root          = strdupz(mi->root);
+        bmi->mount_point   = strdupz(mi->mount_point);
+        bmi->filesystem    = strdupz(mi->filesystem);
+    }
+
+    return bmi;
+}
+
+static void add_basic_mountinfo(struct basic_mountinfo **root, struct mountinfo *mi)
+{
+    if (!root)
+        return;
+
+    struct basic_mountinfo *bmi = basic_mountinfo_create_and_copy(mi);
+
+    bmi->next = *root;
+    *root = bmi;
+};
+
+static void free_basic_mountinfo(struct basic_mountinfo *bmi)
+{
+    if (bmi) {
+        freez(bmi->persistent_id);
+        freez(bmi->root);
+        freez(bmi->mount_point);
+        freez(bmi->filesystem);
+    }
+
+    freez(bmi);
```

Comment:
You tested if `bmi` is not null above. I agree that our `freez` will test if `bmi` is test again for `NULL` pointer, but considering the above test, it would be more clear for people that do not know our code completely to have this `freez` inside of the `if`.

Reply:
From the [`free` man page](https://manned.org/free.3)
> If ptr is NULL, no operation is performed.

So, our `freez` doesn't differ from `free` in any way.

Reply:
I agree 100% with this sentence. I was saying that users that do not know our code, could not understand why we test `bmi` to verify a not null value and we clean after this.

Also If we move `freez(bmi)` inside the first `if` we will avoid a second test for `bmi` when we already detect that the value is NULL.

Reply:
I don't like it. The idea behind checking `bmi` for NULL is different than avoiding another check in `free()`. But let it be - this is a minor thing. Fixed.